src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -77,6 +102,72 @@ class Config:
         Bool('managed_nfsd', default=True),
     )
 
+    @private
+    def setup_directories(self):
+        '''
+        We are moving the NFS state directory from /var/lib/nfs to
+        the system dataset: /var/db/system/nfs.
+        When setup_directories is called /var/db/system/nfs is expected to exist.
+
+        If STATEDIR is empty, then this might be an initialization
+        and there might be current info in /var/lib/nfs.
+
+        We always make sure the expected directories are present
+        '''
+        def name_to_id_conversion(usrgrp):
+            ''' Accept [usr,grp], return [usr,grp] as IDs '''
+            if any(isinstance(owner, str) for owner in usrgrp):
+                # Convert from name to ID, then apply
+                try:
+                    usrgrp[0] = self.middleware.call_sync('user.get_builtin_user_id', usrgrp[0])
+                    usrgrp[1] = self.middleware.call_sync('group.get_builtin_group_id', usrgrp[1])
+                except Exception as e:
+                    self.logger.error(
+                        "%s: Failed to apply ownership. usr=%r, grp=%r %r", usrgrp[0], usrgrp[1], e
+                    )
+            return usrgrp
+
+        # Initialize the system dataset NFS state directory
+        try:
+            if not any(os.scandir(NFSPath.STATEDIR.path())):
+                # System db is empty, populate it with contents of /var/lib/nfs.
+                # This should be a one-time operation.
+                # Going forward, the system dataset will hold the NFS state data.
+                try:
+                    shutil.copytree('/var/lib/nfs', NFSPath.STATEDIR.path())
+                except FileExistsError:
+                    pass
+                except Exception as e:
+                    self.logger.error('Failed to initialize NFS state from /var/lib/nfs: %r', e)
+                # Continue anyway.
+        except Exception as e:
+            self.logger.error("Could not find required path: %r", e)
+
+        # Make sure we have the necessary directories
+        for p in NFSPath:
```

Comment:
This for loop is prone to a few different race conditions. It's also using the "file handle" instead of opening the file/directory and then using the fd based syscalls. I think we need to simplify this to something simple.
     
Something like:
```
for i in NFSServicePathInfo:
    uid = self.name_to_id_conversion(i.owner()['uid'], name_type='user')
    gid = self.name_to_id_conversion(i.owner()['gid'], name_type='group')
    path = i.path()
    if i.is_dir():
        open_flags = os.O_RDWR | os.O_DIRECTORY
        os.makedirs(path, exist_ok=True)
    else:
        open_flags = os.O_RDWR

    try:
        with open(os.open(path, flags=open_flags)) as f:
            os.fchmod(f.fileno(), i.mode())
            os.fchown(f.fileno(), uid, gid)
    except IsADirectoryError:
        if not i.is_dir():
            # the path is expected to be a file but it's a directory
            self.logger.error('Expected %r to be a file but instead found a directory', path)
        else:
            self.logger.error('Unexpected failure updating file %r', path, exc_info=True)
    except NotADirectoryError:
        if i.is_dir():
            # the path is expected to be a directory but it's a file
            self.logger.error('Expected %r to be a directory but instead found a file', path)
        else:
            self.logger.error('Unexpected failure updating directory %r', path, exc_info=True)
    except Exception:
        self.logger.error('Unexpected failure initializing %r', path, exc_info=True)

Reply:
Implemented the simplified solution.